### PR TITLE
Add Foreign key Cascade engine primitive

### PR DESCRIPTION
**Pull request by harshit-gangal**
## Description

This PR adds an engine primitive that can execute a foreign key cascade.
The planner support to support this use case is still not done.

## Related Issue(s)

- #12967 

## Checklist

-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on the CI

### CODE REVIEW

**Review by GuptaManan100 — APPROVED** [review]

File: go/vt/vtgate/engine/fk_cascade.go [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+	Cols   []int // indexes
+	Exec   Primitive
+}
+
+// FkCascade is a primitive that implements foreign key cascading using Selection as values required to execute the FkChild Primitives.
+// On success, it executes the Parent Primitive.
+type FkCascade struct {
+	Selection Primitive
+	Children  []FkChild
+	Parent    Primitive
```
Comment (systay) [inline go/vt/vtgate/engine/fk_cascade.go:39]: comments plz it's not self-evident what Selection and Parent means

File: go/vt/vtgate/engine/fk_cascade.go [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+}
+
+// GetTableName implements the Primitive interface.
+func (fkc *FkCascade) GetTableName() string {
+	return fkc.Parent.GetTableName()
+}
+
+// GetFields implements the Primitive interface.
+func (fkc *FkCascade) GetFields(_ context.Context, _ VCursor, _ map[string]*querypb.BindVariable) (*sqltypes.Result, error) {
+	return nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "[BUG] GetFields should not be called")
```
Comment (systay) [inline go/vt/vtgate/engine/fk_cascade.go:61]: why not?

File: go/vt/vtgate/engine/fk_cascade.go [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+			for _, colIdx := range child.Cols {
+				tuple.Values = append(tuple.Values,
+					sqltypes.ValueToProto(row[colIdx]))
+			}
+			bv.Values = append(bv.Values, tuple)
+		}
+		// Execute the child primitive, and bail out incase of failure.
+		// Since this Primitive is always executed in a transaction, the changes should
+		// be rolled back incase of an error.
+		bindVars[child.BVName] = bv
```
Comment (systay) [inline go/vt/vtgate/engine/fk_cascade.go:98]: did you consider cloning the bindVars map to separate between the one sent to the cascades and the one sent to the Selection and the Parent?

**Review by systay — APPROVED** [review]

File: go/vt/vtgate/engine/fk_cascade.go (hunk 2 above)
Comment (harshit-gangal) [reply go/vt/vtgate/engine/fk_cascade.go:61]: There is nothing to return on `GetFields` on a DML.
prepare statement ignore calling into `GetFields` for DMLs

File: go/vt/vtgate/engine/fk_cascade.go (hunk 3 above)
Comment (harshit-gangal) [reply go/vt/vtgate/engine/fk_cascade.go:98]: we can do it when we see the need for it, after using it in the plan and checking on what additional information is sent down which can be avoided.

**Pull request merged**